src/backend/tls.rs

```diff
@@ -0,0 +1,247 @@
+                    KeyUsage::new()
+                        .critical()
+                        .digital_signature()
+                        .key_cert_sign()
+                        .crl_sign()
+                        .build()?,
+                ];
+                for ext in exts {
+                    x509.append_extension(ext)?;
+                }
+                x509.sign(&ca_key, MessageDigest::sha256())?;
+                Ok(x509)
+            })
+            .unwrap()
+            .build();
+        Self { ca_key, ca_cert }
+    }
+
+    pub async fn get_or_create_k8s_certificate(
+        client: &stackable_operator::client::Client,
+    ) -> Self {
+        let k8s_secret_name = "secret-provisioner-ca";
+        let k8s_ns = "default";
+        let existing_secret = client.get::<Secret>(k8s_secret_name, Some(k8s_ns)).await;
+        match existing_secret {
+            Ok(ca_secret) => {
+                let ca_data = ca_secret.data.unwrap_or_default();
+                Self {
+                    ca_key: PKey::private_key_from_pem(&ca_data.get("ca.key").unwrap().0).unwrap(),
+                    ca_cert: X509::from_pem(&ca_data.get("ca.crt").unwrap().0).unwrap(),
+                }
+            }
+            Err(_) => {
+                // Failed to get existing cert, try to create a new self-signed one
+                let ca = Self::new_self_signed();
+                // Use create rather than apply so that we crash and retry on conflicts (to avoid creating spurious certs that we throw away immediately)
+                client
+                    .create(&Secret {
+                        metadata: ObjectMetaBuilder::new()
+                            .namespace(k8s_ns)
+                            .name(k8s_secret_name)
+                            .build(),
+                        data: Some(
+                            [
+                                (
+                                    "ca.key".to_string(),
+                                    ByteString(ca.ca_key.private_key_to_pem_pkcs8().unwrap()),
+                                ),
+                                (
+                                    "ca.crt".to_string(),
+                                    ByteString(ca.ca_cert.to_pem().unwrap()),
+                                ),
+                            ]
+                            .into(),
+                        ),
+                        ..Secret::default()
+                    })
+                    .await
+                    .unwrap();
+                ca
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl SecretBackend for TlsGenerate {
+    type Error = Error;
+
+    async fn get_secret_data(
+        &self,
+        selector: super::SecretVolumeSelector,
+        pod_info: PodInfo,
+    ) -> Result<SecretFiles, Self::Error> {
+        let now = OffsetDateTime::now_utc();
+        let not_before = now - Duration::minutes(5);
+        let not_after = now + Duration::days(1);
```

Comment:
I think this should be significantly higher, as most of the services we provision will run for a long time without a restart and do not perform dynamic reloading of certificates.
Ideally this should be configurable by the requestor at some point.

Reply:
We'll need to handle reloading the certificates somehow anyway (whether that be through hot reloading or cold-blooded murder), regardless of whether we make the expiry time 5 minutes or 5 weeks. Agreed that we'll probably need to handle long-lasting certificates *somehow*, but I think that should be the exception rather than the rule.

Reply:
For the large majority of products I think a restart is required to activate new certificates, which can make this an expensive operation. But wherever possible we should make a hot-reload possible.

To mitigate this, could we make the runtime user configurable? Maybe the SecretClass for AutoTls prvisioners could get settings for `defaultCertValidity` and `maximumCertValidity`?
If nothing is specified in the volume default gets pulled and within the other parameter longer validities can be requested by a parameter on the volume.

Reply:
`maxCertValidity` on the `SecretClass` makes sense to me. `defaultCertValidity` feels like a footgun that should be managed by the individual operators instead (I'm a bit terrified of the future medium/stackoverflow answers telling people that they should "just set `defaultCertValidity` to a year or whatever" :P).

Reply:
That works for me.